precompiles/identity/src/lib.rs

```diff
@@ -0,0 +1,657 @@
+// Copyright 2019-2023 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile to receive GMP callbacks and forward to XCM
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use enumflags2::BitFlags;
+use fp_evm::PrecompileHandle;
+use frame_support::dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo};
+use frame_support::sp_runtime::traits::StaticLookup;
+use frame_support::traits::Currency;
+use pallet_evm::AddressMapping;
+use precompile_utils::prelude::*;
+use sp_core::{ConstU32, Get, H160, H256, U256};
+use sp_std::marker::PhantomData;
+
+type BalanceOf<T> = <<T as pallet_identity::Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+/// A precompile to wrap the functionality from pallet-identity
+pub struct IdentityPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+#[precompile::test_concrete_types(mock::Runtime)]
+impl<Runtime> IdentityPrecompile<Runtime>
+where
+	Runtime: pallet_identity::Config + pallet_evm::Config,
+	Runtime::AccountId: Into<H160>,
+	Runtime::Hash: From<H256>,
+	Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+	Runtime::RuntimeCall: From<pallet_identity::Call<Runtime>>,
+	BalanceOf<Runtime>: TryFrom<U256> + Into<U256> + solidity::Codec,
+{
+	#[precompile::public("addRegistrar(address)")]
+	fn add_registrar(handle: &mut impl PrecompileHandle, account: Address) -> EvmResult {
+		let account =
+			Runtime::Lookup::unlookup(Runtime::AddressMapping::into_account_id(account.0));
+		let call = pallet_identity::Call::<Runtime>::add_registrar { account };
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("setIdentity((bool, bytes)[], (bool, bytes), (bool, bytes), (bool, bytes), (bool, bytes), (bool, bytes), bool, bytes, (bool, bytes), (bool, bytes))")]
```

Comment:
:eyes:

Reply:
It would be nice if the derive macro for `IdentityInfo` could generate this for us

Reply:
Yeah :P luckily there's a generated test case which tests this with `cargo test` but I agree it's quite hard to read through this.

Reply:
The macro can by itself generate the signature of the function, but we currently require to also write it manually to ensure there was no mistakes with the types. This is checked by an automatically generated test.